### ISSUE TITLE
Bounding Box Regression Equation

### ISSUE THREAD

**Issue by developer0hye** [ultralytics/yolov5#4373]
## ❔Question
Hi, @glenn-jocher !

I have one question.

![image](https://user-images.githubusercontent.com/35001605/128881880-2730a1dc-a474-4bbd-8b96-e82fa0b6353f.png)

As you know, the bounding box regression equation of yolov5 is different from yolov3 and yolov4. However, it is the same with scaled-yolov4. Why do you replace the equation for predicting the width and height of the bounding box?

It seems that yolov5 and scaled-yolov4's equation are more numerically stable than the old ones especially when we train the model with mixed precision.

**Comment by mxy5201314** [comment]
Get more positive samples, accelerate convergence and be more stable

**Comment by developer0hye** [comment]
@mxy5201314 
Thanks for your reply!
How do we get more positive samples with changing the equation?

**Comment by mxy5201314** [comment]
in yolov5, bx belong to (-0.5 + cx) to (1.5 + cx), ~ , bw belong to 0 to (4 * pw), ~,  In these areas are positive samples, You can compare other versions.

**Comment by developer0hye** [comment]
@mxy5201314 
Okay, I could understand your great explanation! 

We can assign more positive samples during training with changing the equation for localization(bx, by). 

But I couldn't understand why the equation for bw and bh is changed for more positive samples.

**Comment by mxy5201314** [comment]
@developer0hye
In fact, the gradient explosion can be prevented without exp, and the convergence can be accelerated by changing bw. Imagine that if tw is too large when using exp, it may exceed the aspect ratio threshold. If tw is too small (less than 0), its gradient is also very small (because it is exp), which is not conducive to training. I think this change is mainly the optimization that the author thinks.

**Comment by developer0hye** [comment]
@mxy5201314 Thanks! I agree with you. Exp operation can cause gradient explosion!

**Comment by glenn-jocher** [comment]
@developer0hye this is because YOLOv4-scaled is based upon 99% of the YOLOv5 codebase which the authors completely fail to mention. Essentially it's a pirated version of YOLOv5 which does not even follow the legal requirements of our GPL 3.0 license.

**Comment by glenn-jocher** [comment]
@developer0hye @mxy5201314 this is because YOLOv4-scaled is based upon 99% of the YOLOv5 codebase (including my box regression equation that you ask about above). The authors completely fail to mention this as they want to present it as their own work, and they are using YOLOv5 augmentations, loss function, regression equation, autoanchor, etc. with zero citation or credit to myself or Ultralytics.

So the box regression equation above I created myself for YOLOv5 in May 2020, it features increased stability during early training due to it's bounded limits, in addition to leveraging a single Sigmoid output on all neurons rather than exp() on some and sigmoid() on others, creating a simpler architecture for YOLOv5 Detect() layer, and then later on YOLOv4-scaled came around and copied my work, stripping all comments, readme's etc that would mention YOLOv5, which is why my box regression equation is in YOLOv4-scaled now as well.

**Comment by mxy5201314** [comment]
@glenn-jocher I trust you and your yolov5 is great.

**Comment by developer0hye** [comment]
@mxy5201314 I think so... @glenn-jocher  is great.

**Comment by developer0hye** [comment]
@glenn-jocher 

Is a network scaling approach proposed in Scaled-YOLOv4 also yolov5's method?

**Comment by glenn-jocher** [comment]
@developer0hye the compound scaling in YOLOv5 is based very closely on the compound scaling concepts (image size, width, depth) first introduced by EfficientDet https://arxiv.org/abs/1911.09070, though I disconnected the image-size from the other two.

The authors say their scaling concepts were inspired by their earlier work [39] Mingxing Tan and Quoc V. Le. Efficientnet: Rethinking model scaling for convolutional neural networks. ICML, 2019

**Comment by developer0hye** [comment]
@glenn-jocher Thanks for your great explanation!

**Comment by WongKinYiu** [comment]
> The authors completely fail to mention this as they want to present it as their own work, and they are using YOLOv5 augmentations, loss function, regression equation, autoanchor, etc. with zero citation or credit to myself or Ultralytics.

Hello,

We greatly thanks for your wonderful works, and mentioned it in ArXiv, paper, and github... We do not use autoanchor in scaled-YOLOv4 and the loss function is CIoU which is proposed by Zheng *et al*. Because all of these credit are yours (mosaic augmentation, hyper-parameter evolution, bounding box regression equation), we never claim these parts are our contributions.
(Mosaic augmentation, hyper-parameter evolution, bounding box regression equation.)
![image](https://user-images.githubusercontent.com/12152972/129434518-4fc5c783-4775-4340-86aa-2fa2980b2eb1.png)
(About 50% are from darknet, 30% are from yolov3, and 5~10% are from here.)
![image](https://user-images.githubusercontent.com/12152972/129434689-5eccbdfa-8816-4e32-b846-1b2562bca777.png)
(github, corresponding to acknowledgements in paper.)
![image](https://user-images.githubusercontent.com/12152972/129434799-f3086d6f-5868-419e-8887-57ba8f665162.png)
(github, shown in first line of readme.)
![image](https://user-images.githubusercontent.com/12152972/129435041-f01039f2-255a-407d-ac8a-7ee81d16ec64.png)

By the way, I do not know why you say

> ... , and then later on YOLOv4-scaled came around and copied my work, ...

while you actually know we develop yolov4-csp and trained the models by your yolov3 code before this repository public released (May 27, 2020.).

![image](https://user-images.githubusercontent.com/12152972/129435949-6c1af9b3-9717-40a2-ad42-bd0e4e3df80b.png)

Scaled-YOLOv4 is originally developing on darknet and your yolov3 codebase, and due to this codebase is really wonderful and easy to use for everyone, we then later implement scaled-YOLOv4 based on this repository. And we will keep mention your great effort and give those credit to you if we use the functions. In actually, I always put the reference link on my report due to I think a large amount discussion are not only from us but also many githubers. 

![image](https://user-images.githubusercontent.com/12152972/129437487-8518e697-755a-466c-95a1-f9a4c2f195c9.png)
![image](https://user-images.githubusercontent.com/12152972/129437499-6f9912a3-8ad3-44d7-9bdc-cefc411739f4.png)
![image](https://user-images.githubusercontent.com/12152972/129437506-b938ac39-6f65-493b-970a-8fb2443dd96a.png)
![image](https://user-images.githubusercontent.com/12152972/129437521-ac409cb4-f691-4e21-bc26-43bb4023a126.png)

**Comment by Zzh-tju** [comment]
It is very common that people propose several tricks A, B, C based on the model D, and then rename D to a new model E which claims to be SOTA. I also use mmdetection, in which more than 99% source codes are not proposed by me. But it doesn't affect that I raname the model to a new SOTA one. Correct and necessary citations and acknowledgements are enough to express my respect to these researchers.

**Comment by jimmyflycv** [comment]
> @developer0hye @mxy5201314 this is because YOLOv4-scaled is based upon 99% of the YOLOv5 codebase (including my box regression equation that you ask about above). The authors completely fail to mention this as they want to present it as their own work, and they are using YOLOv5 augmentations, loss function, regression equation, autoanchor, etc. with zero citation or credit to myself or Ultralytics.
> 
> So the box regression equation above I created myself for YOLOv5 in May 2020, it features increased stability during early training due to it's bounded limits, in addition to leveraging a single Sigmoid output on all neurons rather than exp() on some and sigmoid() on others, creating a simpler architecture for YOLOv5 Detect() layer, and then later on YOLOv4-scaled came around and copied my work, stripping all comments, readme's etc that would mention YOLOv5, which is why my box regression equation is in YOLOv4-scaled now as well.

Totally agree.
